### DISCUSSION TITLE
What is the message for a node with no connections?

### DISCUSSION THREAD

**Discussion by davidireland-iso** [pyg-team/pytorch_geometric#4796] [Q&A] [answered]
Suppose we have a node with no connections and our GNN uses Message Passing Layers. 

If we consider the general case outlined in the documentation, a Message Passing layer calculates messages by applying some function (typically MLP for instance) to the node features for nodes `i`, `j`, and an edge feature between them, `e_(i,j)`. The input to the MLP is typically these vectors concatenated together. However, if node `i` has no connections then what do we use as input to the MLP. Are the vectors for node `j` and the edge feature initialised to 0?

**Comment by rusty1s** [comment, marked as answer]
If you have no edges, then the output of message will be a tensor of shape `[0, num_features]`. After aggregation, you will obtain an zero-initialized tensor of shape `[num_nodes, num_features]`. As such, most implementations either add a self-loop to the graph or apply a simple skip-connection: `out = self.propagate(...) + self.lin(x)`
